futures-channel/src/mpsc/mod.rs

```diff
@@ -305,6 +304,21 @@ struct State {
     num_messages: usize,
 }
 
+#[derive(Debug)]
+struct BoundedInner<T> {
+    // Max buffer size of the channel.
+    buffer: usize,
+
+    // Same state fields as in unbounded queue.
+    unbounded: UnboundedInner<T>,
+
+    // Atomic, FIFO queue used to send parked task handles to the receiver.
+    parked_queue: Queue<Arc<Mutex<SenderTask>>>,
```

Comment:
Just some further thoughts: This could potentially use either a similar approach or directly the `ManualResetEvent` from my CR to wake-up interested `Sender`s. Those would poll the future for waiting on new capacity and then do do a `try_send()`. The advantage would be that the dynamically allocated `Queue` and the `SenderTask` allocations would go away. 

However after thinking about it again it doesn't really fit the current API contract of the channel, since e.g. `poll_ready` doesn't return a future that can be used to directly wait until capacity is available, but is an instance method on the `Sender`, which isn't suitable for storing the intrusive waiter.

Nevertheless might be an interesting thing to try and benchmark at some point of time.